src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -305,76 +297,116 @@ public async Task StoreAsync(object entity, string id, IMetadataDictionary metad
                         metadata[Constants.Documents.Metadata.RavenClrType] = clrType;
                 }
 
-                if (_first == false)
+                await WriteToStream(() =>
                 {
-                    _currentWriter.Write(',');
-                }
+                    if (_first == false)
+                    {
+                        _currentWriter.Write(',');
+                    }
+
+                    _first = false;
 
-                _first = false;
-                try
-                {
                     _currentWriter.Write("{\"Id\":\"");
-                    WriteId(_currentWriter, id);
+                    WriteString(_currentWriter, id);
                     _currentWriter.Write("\",\"Type\":\"PUT\",\"Document\":");
 
                     if (_customEntitySerializer == null || _customEntitySerializer(entity, metadata, _currentWriter) == false)
                     {
                         using (var json = EntityToBlittable.ConvertEntityToBlittable(entity, _conventions, _context,
-                            _defaultSerializer, new DocumentInfo
-                            {
-                                MetadataInstance = metadata
-                            }))
-
+                            _defaultSerializer, new DocumentInfo { MetadataInstance = metadata }))
                         {
                             _currentWriter.Flush();
                             json.WriteJsonTo(_currentWriter.BaseStream);
                         }
                     }
 
                     _currentWriter.Write("}");
-                    _currentWriter.Flush();
+                }, id, CommandType.PUT).ConfigureAwait(false);
+            }
+        }
 
-                    if (_currentWriter.BaseStream.Position > _maxSizeInBuffer ||
-                        _asyncWrite.IsCompleted)
-                    {
-                        await _asyncWrite.ConfigureAwait(false);
-
-                        var tmp = _currentWriter;
-                        _currentWriter = _backgroundWriter;
-                        _backgroundWriter = tmp;
-                        _currentWriter.BaseStream.SetLength(0);
-                        ((MemoryStream)tmp.BaseStream).TryGetBuffer(out var buffer);
-                        _asyncWrite = _requestBodyStream.WriteAsync(buffer.Array, buffer.Offset, buffer.Count, _token);
-                    }
+        private async Task WriteToStream(Action writeOperation, string documentId, CommandType currentCommand)
+        {
+            if (_continousCommandType != CommandType.None && currentCommand != _continousCommandType)
+                ThrowUnfinishedCommand(currentCommand);
+
+            try
+            {
+                writeOperation();
```

Comment:
I don't particularly love this API. 
This force us to do allocations for the delegate on each call, which we would like to avoid.
Can we avoid this?

Reply:
Done